File: client/src/app/(dashboard)/dashboards/utils/timeIntelligence.ts
/**
 * Shared time-intelligence helpers for dashboards, filters, and KPI PoP.
 * Calendar is Gregorian; optional fiscalYearStartMonth (1–12) for fiscal periods.
 */

import dayjs, { type Dayjs } from 'dayjs';

export type ComparisonPeriod = 'wow' | 'mom' | 'qoq' | 'yoy';

export type TimeIntelligenceOptions = {
  /** Reserved for org timezone; currently uses local browser clock. */
  timeZone?: string | null;
  /** Month number 1–12 when fiscal year starts. Default 1 (calendar year). */
  fiscalYearStartMonth?: number;
};

export const COMPARISON_PERIOD_LABELS: Record<ComparisonPeriod, string> = {
  wow: 'last week',
  mom: 'last month',
  qoq: 'last quarter',
  yoy: 'last year',
};

function nowInZone(_opts?: TimeIntelligenceOptions): Dayjs {
  // Local clock today; wire org TZ when dashboard settings expose it.
  return dayjs();
}

function quarterStartMonth(d: Dayjs, fiscalStart = 1): number {
  const m = d.month(); // 0-based
  const fiscalOffset = ((m - (fiscalStart - 1)) % 12 + 12) % 12;
  const qIndex = Math.floor(fiscalOffset / 3);
  return (fiscalStart - 1 + qIndex * 3) % 12;
}

/** Month-to-date: start of month → today */
export function rangeMtd(opts?: TimeIntelligenceOptions): { from: string; to: string } {
  const n = nowInZone(opts);
  return { from: n.startOf('month').format('YYYY-MM-DD'), to: n.format('YYYY-MM-DD') };
}

/** Quarter-to-date (calendar or fiscal) → today */
export function rangeQtd(opts?: TimeIntelligenceOptions): { from: string; to: string } {
  const n = nowInZone(opts);
  const fiscalStart = opts?.fiscalYearStartMonth ?? 1;
  const qStart = quarterStartMonth(n, fiscalStart);
  const from = n.month(qStart).startOf('month');
  return { from: from.format('YYYY-MM-DD'), to: n.format('YYYY-MM-DD') };
}

/** Year-to-date (calendar or fiscal) → today */
export function rangeYtd(opts?: TimeIntelligenceOptions): { from: string; to: string } {
  const n = nowInZone(opts);
  const fiscalStart = opts?.fiscalYearStartMonth ?? 1;
  if (fiscalStart <= 1) {
    return { from: n.startOf('year').format('YYYY-MM-DD'), to: n.format('YYYY-MM-DD') };
  }
  const month = n.month() + 1;
  let fyStart = n.month(fiscalStart - 1).startOf('month');
  if (month < fiscalStart) {
    fyStart = fyStart.subtract(1, 'year');
  }
  return { from: fyStart.format('YYYY-MM-DD'), to: n.format('YYYY-MM-DD') };
}

/**
 * Shift an inclusive date range backward by a comparison period.
 * Preserves duration for wow (7d) and uses calendar month/quarter/year for others.
 */
export function shiftDateRangeForComparison(
  from: string,
  to: string,
  period: ComparisonPeriod,
): { from: string; to: string } {
  const start = dayjs(from);
  const end = dayjs(to);
  if (!start.isValid() || !end.isValid()) {
    return { from, to };
  }

  switch (period) {
    case 'wow':
      return {
        from: start.subtract(7, 'day').format('YYYY-MM-DD'),
        to: end.subtract(7, 'day').format('YYYY-MM-DD'),
      };
    case 'mom':
      return {
        from: start.subtract(1, 'month').format('YYYY-MM-DD'),
        to: end.subtract(1, 'month').format('YYYY-MM-DD'),
      };
    case 'qoq':
      return {
        from: start.subtract(3, 'month').format('YYYY-MM-DD'),
        to: end.subtract(3, 'month').format('YYYY-MM-DD'),
      };
    case 'yoy':
      return {
        from: start.subtract(1, 'year').format('YYYY-MM-DD'),
        to: end.subtract(1, 'year').format('YYYY-MM-DD'),
      };
    default:
      return { from, to };
  }
}

type FilterLike = {
  field?: string;
  operator?: string;
  value?: unknown;
  [key: string]: unknown;
};

function isDateLikeString(v: unknown): v is string {
  return typeof v === 'string' && /^\d{4}-\d{2}-\d{2}/.test(v.trim());
}

/**
 * Shift >= / <= date filters for period-over-period. Returns null if no date bounds found.
 */
export function shiftFiltersForComparison(
  filters: FilterLike[] | null | undefined,
  period: ComparisonPeriod,
): FilterLike[] | null {
  if (!Array.isArray(filters) || filters.length === 0) return null;

  let from: string | null = null;
  let to: string | null = null;
  let fromField: string | null = null;
  let toField: string | null = null;

  for (const f of filters) {
    const op = String(f.operator || '').trim();
    if ((op === '>=' || op === '>') && isDateLikeString(f.value)) {
      from = String(f.value).slice(0, 10);
      fromField = String(f.field || '');
    }
    if ((op === '<=' || op === '<') && isDateLikeString(f.value)) {
      to = String(f.value).slice(0, 10);
      toField = String(f.field || '');
    }
    if (op === 'between' && Array.isArray(f.value) && f.value.length >= 2) {
      if (isDateLikeString(f.value[0]) && isDateLikeString(f.value[1])) {
        from = String(f.value[0]).slice(0, 10);
        to = String(f.value[1]).slice(0, 10);
        fromField = toField = String(f.field || '');
      }
    }
  }

  if (!from || !to) return null;
  if (fromField && toField && fromField !== toField) return null;

  const shifted = shiftDateRangeForComparison(from, to, period);

  return filters.map((f) => {
    const op = String(f.operator || '').trim();
    if ((op === '>=' || op === '>') && isDateLikeString(f.value)) {
      return { ...f, value: shifted.from };
    }
    if ((op === '<=' || op === '<') && isDateLikeString(f.value)) {
      return { ...f, value: shifted.to };
    }
    if (op === 'between' && Array.isArray(f.value) && f.value.length >= 2) {
      if (isDateLikeString(f.value[0]) && isDateLikeString(f.value[1])) {
        return { ...f, value: [shifted.from, shifted.to] };
      }
    }
    return f;
  });
}

export function isComparisonPeriod(v: unknown): v is ComparisonPeriod {
  return v === 'wow' || v === 'mom' || v === 'qoq' || v === 'yoy';
}

File: server/src/core/time_intelligence.py
"""
Shared time-intelligence helpers for KPI period-over-period and date windows.
Mirrors client dashboards/utils/timeIntelligence.ts (Gregorian calendar).
"""

from __future__ import annotations

from datetime import date, datetime, timedelta
from typing import Any, Dict, List, Optional, Tuple

COMPARISON_PERIOD_LABELS = {
    "wow": "vs last week",
    "mom": "vs last month",
    "qoq": "vs last quarter",
    "yoy": "vs last year",
}

VALID_COMPARISON_PERIODS = frozenset(COMPARISON_PERIOD_LABELS.keys())


def _parse_date(value: Any) -> Optional[date]:
    if value is None:
        return None
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    s = str(value).strip()
    if len(s) >= 10 and s[4] == "-" and s[7] == "-":
        try:
            return date.fromisoformat(s[:10])
        except ValueError:
            return None
    return None


def _add_months(d: date, months: int) -> date:
    """Shift calendar month; clamp day to last valid day of target month."""
    year = d.year + (d.month - 1 + months) // 12
    month = (d.month - 1 + months) % 12 + 1
    # last day of target month
    if month == 12:
        last = date(year + 1, 1, 1) - timedelta(days=1)
    else:
        last = date(year, month + 1, 1) - timedelta(days=1)
    day = min(d.day, last.day)
    return date(year, month, day)


def shift_date_range_for_comparison(
    from_s: str, to_s: str, period: str
) -> Tuple[str, str]:
    start = _parse_date(from_s)
    end = _parse_date(to_s)
    if not start or not end:
        return from_s, to_s

    if period == "wow":
        return (start - timedelta(days=7)).isoformat(), (end - timedelta(days=7)).isoformat()
    if period == "mom":
        return _add_months(start, -1).isoformat(), _add_months(end, -1).isoformat()
    if period == "qoq":
        return _add_months(start, -3).isoformat(), _add_months(end, -3).isoformat()
    if period == "yoy":
        return _add_months(start, -12).isoformat(), _add_months(end, -12).isoformat()
    return from_s, to_s


def _is_date_like(value: Any) -> bool:
    return _parse_date(value) is not None


def shift_filters_for_comparison(
    filters: Optional[List[Dict[str, Any]]], period: str
) -> Optional[List[Dict[str, Any]]]:
    """
    Shift >= / <= / between date filters for PoP.
    Returns None when no date bounds are present or period is invalid.
    """
    if period not in VALID_COMPARISON_PERIODS:
        return None
    if not isinstance(filters, list) or not filters:
        return None

    from_d: Optional[str] = None
    to_d: Optional[str] = None
    from_field: Optional[str] = None
    to_field: Optional[str] = None

    for f in filters:
        if not isinstance(f, dict):
            continue
        op = str(f.get("operator") or "").strip()
        val = f.get("value")
        field = str(f.get("field") or "")
        if op in (">=", ">") and _is_date_like(val):
            from_d = str(val)[:10]
            from_field = field
        if op in ("<=", "<") and _is_date_like(val):
            to_d = str(val)[:10]
            to_field = field
        if op == "between" and isinstance(val, (list, tuple)) and len(val) >= 2:
            if _is_date_like(val[0]) and _is_date_like(val[1]):
                from_d = str(val[0])[:10]
                to_d = str(val[1])[:10]
                from_field = to_field = field

    if not from_d or not to_d:
        return None
    if from_field and to_field and from_field != to_field:
        return None

    shifted_from, shifted_to = shift_date_range_for_comparison(from_d, to_d, period)

    out: List[Dict[str, Any]] = []
    for f in filters:
        if not isinstance(f, dict):
            out.append(f)
            continue
        nf = dict(f)
        op = str(nf.get("operator") or "").strip()
        val = nf.get("value")
        if op in (">=", ">") and _is_date_like(val):
            nf["value"] = shifted_from
        elif op in ("<=", "<") and _is_date_like(val):
            nf["value"] = shifted_to
        elif op == "between" and isinstance(val, (list, tuple)) and len(val) >= 2:
            if _is_date_like(val[0]) and _is_date_like(val[1]):
                nf["value"] = [shifted_from, shifted_to]
        out.append(nf)
    return out
